### PR TITLE
Moving tsv-utils from bioconda

### PR DESCRIPTION
**Pull request by corneliusroemer**
It'd be great if someone @conda-forge/staged-recipes could help me get this D package compiling successfully.

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by corneliusroemer** [comment]
@conda-forge/staged-recipes could you please help me resolve this problem:
```
The reported errors are:
- Encountered problems while solving:
-   - nothing provides requested ldc_linux-64
- 
```

How can I add the ldc compiler as a dependency so that it's available for the build?

**Comment by corneliusroemer** [comment]
Now getting the following error, no clue why
```
Downloading https://github.com/eBay/tsv-utils/archive/refs/tags/v2.2.1.tar.gz/
Traceback (most recent call last):
  File "/home/conda/staged-recipes-copy/.ci_support/build_all.py", line 198, in <module>
    build_all(os.path.join(root_dir, "recipes"), args.arch)
  File "/home/conda/staged-recipes-copy/.ci_support/build_all.py", line 101, in build_all
    build_folders(recipes_dir, folders, arch, channel_urls)
  File "/home/conda/staged-recipes-copy/.ci_support/build_all.py", line 157, in build_folders
    conda_build.api.build([recipe], config=get_config(arch, channel_urls))
  File "/opt/conda/lib/python3.8/site-packages/conda_build/api.py", line 186, in build
    return build_tree(
  File "/opt/conda/lib/python3.8/site-packages/conda_build/build.py", line 3083, in build_tree
    packages_from_this = build(metadata, stats,
  File "/opt/conda/lib/python3.8/site-packages/conda_build/build.py", line 2133, in build
    try_download(m, no_download_source=False, raise_error=True)
  File "/opt/conda/lib/python3.8/site-packages/conda_build/render.py", line 649, in try_download
    source.provide(metadata)
  File "/opt/conda/lib/python3.8/site-packages/conda_build/source.py", line 859, in provide
    unpack(source_dict, src_dir, metadata.config.src_cache, recipe_path=metadata.path,
  File "/opt/conda/lib/python3.8/site-packages/conda_build/source.py", line 139, in unpack
    src_path, unhashed_fn = download_to_cache(cache_folder, recipe_path, source_dict, verbose)
  File "/opt/conda/lib/python3.8/site-packages/conda_build/source.py", line 82, in download_to_cache
    download(url, path)
  File "/opt/conda/lib/python3.8/site-packages/conda/exports.py", line 404, in download
    return _download(url, dst_path, md5=md5sum, sha256=sha256, size=size)
  File "/opt/conda/lib/python3.8/site-packages/conda/common/io.py", line 88, in decorated
    return f(*args, **kwds)
  File "/opt/conda/lib/python3.8/site-packages/conda/gateways/connection/download.py", line 64, in download
    with open(target_full_path, 'wb') as fh:
IsADirectoryError: [Errno 21] Is a directory: '/home/conda/staged-recipes/build_artifacts/src_cache/'
##[error]Bash exited with code '1'.
Finishing: Run docker build
```

**Comment by corneliusroemer** [comment]
@conda-forge-admin please ping as this is ready for review

**Comment by corneliusroemer** [comment]
@conda-forge/staged-recipes ping for review

**Comment by wolfv** [comment]
Did you try to compile this with `ldc` which is available on conda-forge?

**Comment by corneliusroemer** [comment]
@wolfv Yes, tried with ldc, didn't work: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=422874&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=807

Tried with dmd, but dmd on conda-forge is Linux only (has not been maintained properly, still on a release from 2014 - should probably be thrown out unless a maintainer is found and extended to other OSes, for which releases do exist.

**Comment by wolfv** [comment]
That error sounds like you are missing `{{ compiler('c') }}` in that case. Failed to locate `cc` seems to want a `C` compiler, and the env `CC=gcc-..` variable. That variable should be set when you add the `{{ compiler('c') }}`.

**Comment by corneliusroemer** [comment]
@wolfv it now fails with the following error:
```
Error: The LLVMgold.so plugin (needed for LTO) was not found. You can specify its path with -flto-binary=<file>.
```
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=422917&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=954

File: recipes/tsv-utils/build.sh [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+set -x
+
+make DCOMPILER=ldc2 LDC_LTO_RUNTIME=1 LDC_PGO=2
```
Comment (wolfv) [inline recipes/tsv-utils/build.sh:5]: I think if you remove `LDC_PGO` you might not have problems wtih `gold` anymore.

**Comment by corneliusroemer** [comment]
Removing that argument sadly didn't help

```
Error: The LLVMgold.so plugin (needed for LTO) was not found. You can specify its path with -flto-binary=<file>.
```
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=422923&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=949

File: recipes/tsv-utils/build.sh (hunk 1 above)
Comment (corneliusroemer) [reply recipes/tsv-utils/build.sh]: Sadly didn't help
```
Error: The LLVMgold.so plugin (needed for LTO) was not found. You can specify its path with -flto-binary=<file>.
```
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=422923&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=949

Comment (corneliusroemer) [reply recipes/tsv-utils/build.sh]: For mac the error type is different:
```
make -C csv2tsv 
make[1]: Entering directory '$SRC_DIR/csv2tsv'
dyld: lazy symbol binding failed: Symbol not found: _dyld_enumerate_tlv_storage
  Referenced from: /Users/runner/mambaforge/conda-bld/tsv-utils_1638701452887/_build_env/bin/ldc2
  Expected in: /usr/lib/libSystem.B.dylib

dyld: Symbol not found: _dyld_enumerate_tlv_storage
  Referenced from: /Users/runner/mambaforge/conda-bld/tsv-utils_1638701452887/_build_env/bin/ldc2
  Expected in: /usr/lib/libSystem.B.dylib
```
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=422923&view=logs&j=e35d4f76-8ff2-5536-d795-df91e63eb9f7&t=fa7b4b17-b6ff-5c9c-8cfc-15f888c92310&l=1682

Are binaries not accepted at conda-forge? That was working already.

If we can't get it to work here and binaries are not accepted, bioconda accepts binaries. Maybe that's more suitable then?

**Comment by wolfv** [comment]
And if you also remove the LTO runtime stuff?

**Comment by corneliusroemer** [comment]
Trying it out, but then it won't run as fast ;)

**Comment by corneliusroemer** [comment]
@wolfv getting further on Linux, yet still failing. On Mac, still same error.

```
  ERROR (tsv-utils,bin/tsv-summarize): $RPATH/libphobos2-ldc-shared.so.90 not found in packages, sysroot(s) nor the missing_dso_whitelist.
```
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=422936&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=1295

Is the hassle worth it over just using the binaries that are provided in the repo? I had that working already.

**Comment by wolfv** [comment]
That just means that you also need to require the ldc compiler at runtime.

We should at some point split the packages and put runtime packages (phobos, etc.) into a seperate package from ldc.

**Comment by beckermr** [comment]
The RPATHS appear wrong here: `$RPATH/libphobos2-ldc-shared.so.90`. Once that is fixed, which should be easy to change in the makefile or w/e, then this should work fine.

**Comment by corneliusroemer** [comment]
Thanks @beckermr!

Will this also work on Mac? Right now it doesn't build there.

And will it work with LTO again?

image

**Comment by beckermr** [comment]
Linux works now. I couldn't get conda-build to not fail the linking checks, so I added the relevant entries to an allowed list. 

As for OSX, I have no idea, but it appears the build is looking at the wrong system library. It needs to be directed at the one in the SDK we download. IDK how to do this with the D compiler.

**Comment by corneliusroemer** [comment]
That's great, thanks @beckermr. Apart from Ma, it'd be great if LTO was added back, though I don't know how much of a difference that makes for speed.

**Comment by beckermr** [comment]
IDK what the implications of adding back LTO are precisely.

**Comment by beckermr** [comment]
Thanks for the help @isuru. I couldn't find the compiler flags in the docs. :/

It looks like this was added to bioconda so unless the author wants to move it, we should close this.

**Comment by corneliusroemer** [comment]
@beckermr this should really go to conda-forge I presume because it is not bio related.

Reason for submitting to bioconda was that I wasn't sure whether we'd get it to work here and bioconda accepts binaries.

So I'm supportive of moving, if we get it to work here. I just don't know how to fix it, so it's up to you and @isuruf & co if you feel like it's worth the effort.

**Pull request closed without merging**